src/main/java/com/bench/stockmanagement/ProductHandler.java

```diff
@@ -1,35 +1,122 @@
+                .log()
+                .map(productMapper::mapStockData);
+    }
+
+    // Get all information about a product by itemNumber
+    // Name, quantity, price, all orders, all selling
+    public Mono<Product> getProduct(String itemNumber) {
+        SdkPublisher<Page<DBOrder>> ordersByItemNumber = orderRepository.getOrdersByItemNumber(itemNumber);
+        SdkPublisher<Page<DBReceipt>> soldItemsByItemNumber = sellingRepository.getSoldItemsByItemNumber(itemNumber);
+
+        Flux<DBOrder> ordersFlux = Flux.from(ordersByItemNumber)
+                .log()
+                .map(Page::items)
+                .flatMap(Flux::fromIterable);
+
+        Flux<DBReceipt> receiptsFlux = Flux.from(soldItemsByItemNumber)
+                .log()
+                .map(Page::items)
+                .flatMap(Flux::fromIterable);
+
+        return Mono.zip(ordersFlux.collectList(),
+                        receiptsFlux.collectList()).log()
+                .map(objects -> productMapper.map(itemNumber, objects.getT1(), objects.getT2()));
     }
 
-    public Product getProduct(String itemNumber) {
-        List<DBOrder> orderedProducts = orderService.getOrderedProductsByItemNumber(itemNumber);
-        List<DBReceipt> soldItems = sellingService.getSoldItemsByItemNumber(itemNumber);
-        return productMapper.map(itemNumber, orderedProducts, soldItems);
+    // Update the purchase price, stock information
+    public Mono<Result> updateProduct(ProductStockData productStockData) {
+        DBProduct dbProduct = productMapper.mapProduct(productStockData);
+        return Mono.fromFuture(productRepository.getProductByItemNumber(productStockData.getItemNumber()))
+                .doOnSuccess(Objects::requireNonNull)
+                .doOnNext(product -> productRepository.updateProduct(mergeDbProducts(product, dbProduct)))
+                .doOnSuccess(Objects::requireNonNull)
+                .thenReturn(SUCCESS)
+                .onErrorReturn(FAIL);
     }
 
+    private DBProduct mergeDbProducts(DBProduct oldProduct, DBProduct newProduct) {
+        String itemNumber = oldProduct.getItemNumber();
+
+        Integer lastSellingPrice = Optional.ofNullable(newProduct.getLastSellingPrice())
+                .orElse(oldProduct.getLastSellingPrice());
+
+        Map<String, Double> costs = oldProduct.getCosts();
+        if (newProduct.getCosts() != null) {
+            costs.putAll(newProduct.getCosts());
+        }
+
+        Map<String, Integer> minStock = getStockMap(oldProduct.getMinStock(), newProduct.getMinStock());
+        Map<String, Integer> maxStock = getStockMap(oldProduct.getMaxStock(),newProduct.getMaxStock());
+
+        return new DBProduct(itemNumber, lastSellingPrice, minStock, maxStock, costs);
+    }
+
+    private Map<String, Integer> getStockMap(Map<String, Integer> oldStock, Map<String, Integer> newStock) {
+        if (newStock != null) {
+            oldStock.putAll(newStock);
```

Comment:
same here, manipulating the existing `Map` instead of creating a new one can have unwanted side effects
in this particular case, as this method is called twice after each other, this could make the logic incorrect